Add reduceapt function

cycool29 opened the pull request:
Added `reduceapt` function in every app's install & uninstall script that using `apt` to remove unwanted lines from `apt` output.
Added in `purge-installed` script too.

Itai-Nelken commented:
Why not make pi-apps override the apt command? It would be simpler.

cycool29 commented:
> Why not make pi-apps override the apt command? It would be simpler.

You mean like how pi-apps convert `wget` to `aria2c` ? 
I don't really understand how pi-apps convert `wget` to `aria2c`, if there is a way to do so on `apt` command, that will be great !
Maybe something like this:
```
apt() {
	actions = "$*"
	( sudo apt $actions ) 2>&1 | reduceapt

```

However, I think directly modifying the scripts will be more straight-forward way.

Botspot commented:
I've been gradually transitioning all apps from `apt` to the `pkg-install` script. Basically, if an app fails due to an apt error, I transition it.
@cycool29, Looks like you found all the ones I haven't transitioned yet!

Using reduceapt by itself will only **reduce** apt's output. That won't **diagnose** apt-errors for the user, which is the important thing.
By the way, unless I'm mistaken, none of these scripts will work, because the `reduceapt()` function is not in the `api` script, it's in the `pkg-install` script.

cycool29 commented:
> I've been gradually transitioning all apps from apt to the pkg-install script. Basically, if an app fails due to an apt error, I transition it.
@cycool29, Looks like you found all the ones I haven't transitioned yet!

Lol, I just hate seeing those `WARNING: apt does not have a stable cli interface. Use with caution in scripts` and I just search in the repo, append `reduceapt` to all of them .

> By the way, unless I'm mistaken, none of these scripts will work, because the reduceapt() function is not in the api script, it's in the pkg-install script. (though that's easy to change)

Well, I have defined `reduceapt` function in those script.

Botspot commented:
> Lol, I just hate seeing those `WARNING: apt does not have a stable cli interface. Use with caution in scripts` and I just search in the repo, append `reduceapt` to all of them .

That warning appears because Pi-Apps appends the script's output to a logfile, and to the terminal. APT thinks stdout is not a terminal. @cycool29, if you know how to solve this (how to make apt think stdout is a terminal), that would be great.

cycool29 commented:
@Botspot I think there should be a `pkg-purge` script, just like `pkg-install`, but just to purge packages in installation script to avoid conflicts. ( like removing `steamlink` when installing Steam because it conflicts with steam-launcher )

>  @cycool29, if you know how to solve this (how to make apt think stdout is a terminal), that would be great.

I am searching a whole day for this and I found `unbuffer` seems able to do that. You can try it by comparing `sudo apt update | grep -v abc` and `unbuffer sudo apt update | grep -v abc`. Make sure you have `expect` installed: `sudo apt install expect`.

Botspot commented:
> I found `unbuffer` seems able to do that. Make sure you have `expect` installed: `sudo apt install expect`.

Interesting. That worked for me. Now the question becomes, "Do we really want to another dependency to Pi-Apps for *this*?"

cycool29 commented:
> Interesting. That worked for me. Now the question becomes, "Do we really want to another dependency to Pi-Apps for this?"

That's why, if that is capable with `reduceapt` (grep -v), there should not be another dependency to pi-apps.
Hence directly add `reduceapt` to the install & uninstall script should be the more straight-forward and effective way.

Maybe add `reduceapt` function in `api` script and append these text in the [Create app wiki](https://github.com/Botspot/pi-apps/wiki/Creating-an-app):

> If you are using `apt` in your script for some reason, add ` | reduceapt` at the end of your line. This will remove the unwanted lines from `apt`.
If you are using `sudo apt purge {package}`, use `sudo apt purge {package} | reduceapt` instead.
If you want to have a error message, use `sudo apt purge {package} | reduceapt || error "Your error message here"`

Jai-JAP commented:
Adding `apt` function and `reduceapt` function to api will work.

```
reduceapt() { #remove unwanted lines from apt output
  grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
}
    
apt() {
    if [ $(id -u) -ne 0 ]; then
      sudo -E apt $@ 2>&1| reduceapt
    else
      "apt" $@ 2>&1 | reduceapt
    fi
}
```

Jai-JAP commented:
But using `pkg-install` is the best option as it also diagnoses apt errors
(Moreover It has `reduceapt` lol)

cycool29 commented:
> But using `pkg-install` is the best option as it also diagnoses apt errors (Moreover It has `reduceapt` lol)

+1.
So just change all the `apt install` to `pkg-install` and add `reduceapt` function only in `apt purge` script.

cycool29 commented:
> @Botspot I think there should be a pkg-purge script, just like pkg-install, but just to purge packages in installation script to avoid conflicts. ( like removing steamlink when installing Steam because it conflicts with steam-launcher )

Botspot commented:
> > @Botspot I think there should be a pkg-purge script, just like pkg-install, but just to purge packages in installation script to avoid conflicts. ( like removing steamlink when installing Steam because it conflicts with steam-launcher )
> 
> If this is possible, we can just add `reduceapt` in `pkg-purge` and replace all `apt purge` to `pkg-purge`

Or... simply suppress all output like QEMU does.
```
#Remove any qemu packages first to prevent conflicting packages
sudo apt purge -yqq qemu* &>/dev/null
```

cycool29 commented:
> > > @Botspot I think there should be a pkg-purge script, just like pkg-install, but just to purge packages in installation script to avoid conflicts. ( like removing steamlink when installing Steam because it conflicts with steam-launcher )
> > 
> > 
> > If this is possible, we can just add `reduceapt` in `pkg-purge` and replace all `apt purge` to `pkg-purge`
> 
> Or... simply suppress all output like QEMU does.
> 
> ```
> #Remove any qemu packages first to prevent conflicting packages
> sudo apt purge -yqq qemu* &>/dev/null
> ```

It is an option too, but that will be hard to diagnose if there is any error during uninstallation.

cycool29 commented:
Well, I have updated all scripts using `apt install` to use `pkg-install`. 
Now left `apt purge`, should we remove all output from `apt purge` or just remove unwanted lines?

Jai-JAP commented:
```
reduceapt() { #remove unwanted lines from apt output
  grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
}

pkg-purge() {
    purge_list="$1" #quotation-marked, space-seperated list of package names to install
    app=$(basename "$2")
    echo $purge_list > "${2}/../../data/purged-packages/${app}"
    if [ $(id -u) -ne 0 ]; then 
      sudo -E apt purge $1 2>&1| reduceapt
    else
      apt purge $1 2>&1 | reduceapt
    fi
}

restore-purged() {
    app=$(basename "$2")
    if [[ -f "${2}/../../data/purged-packages/${app}" -a -n $(cat "${2}/../../data/purged-packages/${app}") ]] ; then
      if [ $(id -u) -ne 0 ]; then 
        sudo -E apt install $1 2>&1| reduceapt
      else
        apt install $1 2>&1| reduceapt
      fi
    fi
}
```


Syntax
pkg-purge "package1 package2" "${DIRECTORY}"

cycool29 commented:
> ```
> reduceapt() { #remove unwanted lines from apt output
>   grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
> }
> 
> pkg-purge() {
>     purge_list="$1" #quotation-marked, space-seperated list of package names to install
>     app=$(basename "$2")
>     echo $purge_list > "${2}/../../data/purged-packages/${app}"
>     if [ $(id -u) -ne 0 ]; then 
>       sudo -E apt purge $1 2>&1| reduceapt
>     else
>       apt purge $1 2>&1 | reduceapt
>     fi
> }
> 
> restore-purged() {
>     app=$(basename "$2")
>     if [[ -f "${2}/../../data/purged-packages/${app}" -a -n $(cat "${2}/../../data/purged-packages/${app}") ]] ; then
>       if [ $(id -u) -ne 0 ]; then 
>         sudo -E apt install $1 2>&1| reduceapt
>       else
>         apt install $1 2>&1| reduceapt
>       fi
>     fi
> }
> ```
> 
> Syntax: pkg-purge "package1 package2" "${DIRECTORY}" 
> restore-purged "${DIRECTORY}"

The code works for me!

Btw, I think the `pkg-purge` and `restore-purged` should be written as a script instead of a function as `pkg-install` and `purge-installed` is written in script and it may make confusion to app creators.

Jai-JAP commented:
The above is only the code we can add it either to a script or a function after modifying as required

cycool29 commented:
`pkg-purge`
```
#!/bin/bash
{ #prevents errors if script was modified while in use
set -o pipefail #if any command within a pipe fails, make the whole pipe fail. Necessary for error-catching with reduceapt
DIRECTORY="$(readlink -f "$(dirname "$0")")"
trap "exit 1" TERM
export TOP_PID=$$

error() {
  echo -e "\e[91m$1\e[39m" 1>&2
  kill -w -s TERM $TOP_PID
  exit 1
}

PURGE_LIST="$1" #quotation-marked, space-seperated list of package names to install
app="$(basename "$2")" #remove any slashes to just get program name

reduceapt() { #remove unwanted lines from apt output
  grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
}

echo "Running pkg-purge" 
if [ -z "$PURGE_LIST" ];then
  error "No package names specified to pkg-purge!"
elif [ -z "$app" ];then
  error "No app name specified to pkg-purge!"
fi


source "${DIRECTORY}/api"
apt_lock_wait

 
output="$(sudo -E LANG=C LC_ALL=C LANGUAGE=C apt purge $1 --autoremove -y 2>&1 | reduceapt)"
exitcode=$?

errors="$(echo "$output" | grep '^[(W)|(E)|(Err]:')"
if [ $exitcode != 0 ] || [ ! -z "$errors" ];then
  echo -e "\e[91mFailed to uninstall the packages!\e[39m"
  echo -e "APT reported these errors:\n\e[91m$errors\e[39m"
  exit 1
else
  echo $PURGE_LIST > "${DIRECTORY}/data/purged-packages/${app}"
fi 
}

```

Jai-JAP commented:
`restore-purged`

```
#!/bin/bash
{ #prevents errors if script was modified while in use
set -o pipefail #if any command within a pipe fails, make the whole pipe fail. Necessary for error-catching with reduceapt
DIRECTORY="$(readlink -f "$(dirname "$0")")"
trap "exit 1" TERM
export TOP_PID=$$

error() {
  echo -e "\e[91m$1\e[39m" 1>&2
  kill -w -s TERM $TOP_PID
  exit 1
}

app="$(basename "$1")" #remove any slashes to just get program name
PURGED_LIST="$(cat "${DIRECTORY}/data/purged-packages/${app}")" #get the list of purged-packages for $app

reduceapt() { #remove unwanted lines from apt output
  grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
}

echo "Running restore-purged" 
if [ ! -f "${DIRECTORY}/data/purged-packages/${app}" ]; then
  error "purged-packages data file not found for ${app}"
elif [ -z "$PURGED_LIST" ]; then
  error "No packages found to restore-purged!"
elif [ -z "$app" ];then
  error "No app name specified to restore-purged!"
fi


source "${DIRECTORY}/api"
apt_lock_wait

 
output="$(sudo -E LANG=C LC_ALL=C LANGUAGE=C apt install $PURGED_LIST -y 2>&1 | reduceapt)"
exitcode=$?

errors="$(echo "$output" | grep '^[(W)|(E)|(Err]:')"
if [ $exitcode != 0 ] || [ ! -z "$errors" ];then
  echo -e "\e[91mFailed to install the packages!\e[39m"
  echo -e "APT reported these errors:\n\e[91m$errors\e[39m"
  exit 1
else
  rm -f "${DIRECTORY}/data/purged-packages/${app}"
fi 
}
```

cycool29 commented:
I have added `pkg-purge` and replaced all app scripts' `apt` with either `pkg-install` or `pkg-purge` **except** `Downgrade Chromium` as it uses `sudo apt upgrade` and it can be replace with neither `pkg-install` nor `pkg-purge`.

But for `restore-purged`, I am still considering because:
- We don't know if user still need the app
- Make confusions to users (e.g. `Box8`6 in pi-apps detected an `Box86` originally installed in system when running the installation script so removed it with `pkg-purge`, but when uninstallation it purged pi-apps `Box86` **but** reinstalled the original `Box86` )

Jai-JAP commented:
We can add a remember purged packages option to `pkg-purge`

Add `remember_purged=$2`
Change `app=$3`
Replace
```
echo $PURGE_LIST > ${DIRECTORY}/data/purged_packages/${app}
```
with
```
if [[ $remember_purged == 1 ]]; then
  echo $PURGE_LIST > ${DIRECTORY}/data/purged_packages/${app}
fi
```

so that pkg-purge can be used in uninstall scripts without saving `purged_packages`

Jai-JAP commented on apps/Angry IP scanner/uninstall:
```diff
@@ -11,7 +11,7 @@ reduceapt() { #remove unwanted lines from apt output
 }
 
-sudo apt purge -y ipscan 2>&1 | reduceapt
+"${DIRECTORY}/pkg-purge" "ipscan" "$(dirname "$0")" || error "Failed to purge Angry IP Scanner!"
```
This line is not needed as purge-installed will do the work here
This line can be removed

cycool29 replied:
Oops, I forgot to remove it.

Jai-JAP commented:
Also `restore-purge` must be added in Firefox Rapid Release uninstall script as `firefox-esr` must be restored if installed previously

Jai-JAP commented on restore-purged:
```diff
@@ -0,0 +1,46 @@
+  grep -v "apt does not have a stable CLI interface.\|Reading package lists...\|Building dependency tree\|Reading state information...\|Need to get\|After this operation,\|Get:\|Fetched\|Selecting previously unselected package\|Preparing to unpack\|Unpacking \|Setting up \|Processing triggers for "
+}
+
+echo "Running restore-purged" 
+if [ ! -f "${DIRECTORY}/data/purged-packages/${app}" ]; then
+  error "purged-packages data file not found for ${app}"
+elif [ -z "$PURGED_LIST" ]; then
+  error "No packages found to restore-purged!"
+elif [ -z "$app" ];then
+  error "No app name specified to restore-purged!"
+fi
```
Replace error with echo as script must display an error and go on without stopping in above cases
Also add red color to error messages to echo error in red but not stop

Itai-Nelken replied:
red: `\e[1m` normal: `\e[0m`

cycool29 commented:
@Botspot @Itai-Nelken What do you think?

Itai-Nelken commented on apps/Downgrade Chromium/uninstall:
```diff
@@ -7,6 +7,10 @@ function error {
   exit 1
 }
 
+reduceapt() { #remove unwanted lines from apt output
```
There is no need to declare `reduceapt` here, as `api` functions are available to the scripts declaring `reduceapt` in the `api` script makes more sense.

Itai-Nelken commented on pkg-purge:
```diff
@@ -0,0 +1,45 @@
+PURGE_LIST="$1" #quotation-marked, space-seperated list of package names to install
+app="$(basename "$2")" #remove any slashes to just get program name
+
+reduceapt() { #remove unwanted lines from apt output
```
I think it's better to declare `reduceapt` only once in `api`. that makes changing it much easier.

Itai-Nelken commented on restore-purged:
```diff
@@ -0,0 +1,46 @@
+app="$(basename "$1")" #remove any slashes to just get program name
+PURGED_LIST="$(cat "${DIRECTORY}/data/purged-packages/${app}")" #get the list of purged-packages for $app
+
+reduceapt() { #remove unwanted lines from apt output
```
same.

Itai-Nelken commented on purge-installed:
```diff
@@ -11,6 +11,10 @@ function error {
   exit 1
 }
 
+reduceapt() { #remove unwanted lines from apt output
```
same

Itai-Nelken reviewed (changes requested):
Declaring `reduceapt` over and over again is a bad idea, every time we wan't to change it we need to hunt down every script that declares it. instead simply declare it in the `api` script. `api` functions are available to install/uninstall scripts. what do you think @Botspot.

Botspot commented:
> Declaring `reduceapt` over and over again is a bad idea, every time we wan't to change it we need to hunt down every script that declares it. instead simply declare it in the `api` script. `api` functions are available to install/uninstall scripts. what do you think @Botspot.

I completely agree. Honestly I have not been monitoring this PR too closely - I've been re-writing large portions of pi-apps, in order to solve this issue in a far more elegant way.

cycool29 commented:
> This PR might not be merged, however. I've been working on a few major improvements and they will address these issues by converting all app-scripts to an abstract scripting model.

As this, closing.
